package com.renprojects.zonixrental.repositories;

import com.renprojects.zonixrental.models.unit;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Page;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;

import jakarta.persistence.LockModeType;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface UnitRepository extends JpaRepository<unit, Long> {
Optional<unit> findByUnitNumber(String unitNumber);
List<unit> findByIsOccupiedFalse();
Page<unit> findByIsOccupiedFalse(Pageable pageable);

@Lock(LockModeType.PESSIMISTIC_WRITE)
@Query("select propertyUnit from unit propertyUnit where propertyUnit.id = :id")
Optional<unit> findByIdForUpdate(@Param("id") Long id);
}
